### ISSUE TITLE
Check string / literal conversion

### ISSUE THREAD

**Issue by Shinigami92** [palantir/tslint#4094]
### Rule Suggestion

**Is your rule for a general problem or is it specific to your development style?**

general

**What does your suggested rule do?**

Warn the developer that he/she can not cast something like `'abc' as 'def'`

**List several examples where your rule could be used**

```ts
1 as 'def'; // Type 'number' cannot be converted to type '"def"'.
'1' as number; // Type 'string' cannot be converted to type 'number'.
true as string; // Type 'boolean' cannot be converted to type 'string'.

'abc' as 'def'; // working????

function f(x: 'abc' | 'def'): void {
	switch (x) {
		case 'abc':
			doSomething();
			break;
		case 'def':
			doSomethingElse();
			break;
	}
}

f('ghi' as 'abc');
```

**Additional context**

typescript will NOT fix this:
https://github.com/Microsoft/TypeScript/issues/26126

**Comment by JoshuaKGoldberg** [comment]
Note: per #4534, this issue will be closed in less than a month if no PR is sent to add the rule. If you _really_ need the rule, custom rules are always an option and can be maintained outside this repo!

Looking back on this issue's historical context, it was only really useful because TypeScript didn't have `as const` assertions - which meant folks would need to `'abc' as 'abc'`. Now that those exist this rule would actually be more useful IMO.
